File: GKPageScrollView/objc/GKPageSmoothView.m
//
//  GKPageSmoothView.m
//  GKPageScrollViewObjc
//
//  Created by QuintGao on 2020/5/4.
//  Copyright © 2020 gaokun. All rights reserved.
//

#import "GKPageSmoothView.h"
#import "GKPageDefine.h"

static NSString *const GKPageSmoothViewCellID = @"smoothViewCell";

@interface GKPageSmoothCollectionView : UICollectionView<UIGestureRecognizerDelegate>

@property (nonatomic, weak) UIView *headerContainerView;

@end

@implementation GKPageSmoothCollectionView

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldReceiveTouch:(UITouch *)touch {
    CGPoint point = [touch locationInView:self.headerContainerView];
    if (CGRectContainsPoint(self.headerContainerView.bounds, point)) {
        return NO;
    }
    return YES;
}

@end


@interface GKPageSmoothView()<UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout>

@property (nonatomic, weak) id<GKPageSmoothViewDelegate> delegate;
@property (nonatomic, strong) GKPageSmoothCollectionView  *listCollectionView;
@property (nonatomic, strong) NSMutableDictionary <NSNumber *, id<GKPageSmoothListViewDelegate>> *listDict;
@property (nonatomic, strong) NSMutableDictionary <NSNumber *, UIView *> *listHeaderDict;

@property (nonatomic, strong) UIView *headerContainerView;
@property (nonatomic, weak) UIScrollView *currentListScrollView;

@property (nonatomic, assign) BOOL syncListContentOffsetEnabled;
@property (nonatomic, assign) CGFloat currentHeaderContainerViewY;

@property (nonatomic, assign) CGFloat headerContainerHeight;
@property (nonatomic, assign) CGFloat headerHeight;
@property (nonatomic, assign) CGFloat segmentedHeight;
@property (nonatomic, assign) CGFloat currentListInitializeContentOffsetY;

@property (nonatomic, assign) NSInteger currentIndex;

@property (nonatomic, assign) BOOL      isLoaded;

@end

@implementation GKPageSmoothView

- (instancetype)initWithDelegate:(id<GKPageSmoothViewDelegate>)delegate {
    if (self = [super initWithFrame:CGRectZero]) {
        self.delegate = delegate;
        _listDict = [NSMutableDictionary dictionary];
        _listHeaderDict = [NSMutableDictionary dictionary];
        _ceilPointHeight = GKPAGE_NAVBAR_HEIGHT;
        
        [self initSubviews];
    }
    return self;
}

- (void)initSubviews {
    self.headerContainerView = [UIView new];
    
    UICollectionViewFlowLayout *layout = [UICollectionViewFlowLayout new];
    layout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    layout.minimumLineSpacing = 0;
    layout.minimumInteritemSpacing = 0;
    self.listCollectionView = [[GKPageSmoothCollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:layout];
    self.listCollectionView.dataSource = self;
    self.listCollectionView.delegate = self;
    self.listCollectionView.pagingEnabled = YES;
    self.listCollectionView.bounces = NO;
    self.listCollectionView.showsHorizontalScrollIndicator = NO;
    self.listCollectionView.scrollsToTop = NO;
    [self.listCollectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:GKPageSmoothViewCellID];
    if (@available(iOS 11.0, *)) {
        self.listCollectionView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    }
    if (@available(iOS 10.0, *)) {
        self.listCollectionView.prefetchingEnabled = NO;
        
    }
    [self addSubview:self.listCollectionView];
    _listCollectionView.headerContainerView = self.headerContainerView;
    
    [self addSubview:self.headerContainerView];
    [self refreshHeaderView];
}

- (void)dealloc {
    for (id<GKPageSmoothListViewDelegate> listItem in self.listDict.allValues) {
        [listItem.listScrollView removeObserver:self forKeyPath:@"contentOffset"];
    }
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    self.listCollectionView.frame = self.bounds;
}

- (void)refreshHeaderView {
    // 重新设置headerView及segmentedView的frame
    UIView *headerView = [self.delegate headerViewInSmoothView:self];
    UIView *segmentedView = [self.delegate segmentedViewInSmoothView:self];
    [self.headerContainerView addSubview:headerView];
    [self.headerContainerView addSubview:segmentedView];
    
    self.headerHeight = headerView.bounds.size.height;
    self.segmentedHeight = segmentedView.bounds.size.height;
    self.headerContainerHeight = self.headerHeight + self.segmentedHeight;
    
    __weak __typeof(self) weakSelf = self;
    [self refreshWidthCompletion:^(CGFloat width) {
        __strong __typeof(self) self = weakSelf;
        
        self.headerContainerView.frame = CGRectMake(0, 0, width, self.headerContainerHeight);
        headerView.frame = CGRectMake(0, 0, width, self.headerHeight);
        segmentedView.frame = CGRectMake(0, self.headerHeight, width, self.segmentedHeight);
    }];
}

- (void)reloadData {
    self.currentListScrollView = nil;
    self.currentIndex = self.defaultSelectedIndex;
    self.syncListContentOffsetEnabled = NO;
    self.currentHeaderContainerViewY = 0;
    self.isLoaded = YES;
    
    [self.listHeaderDict removeAllObjects];
    
    for (id<GKPageSmoothListViewDelegate> list in self.listDict.allValues) {
        [list.listScrollView removeObserver:self forKeyPath:@"contentOffset"];
        [list.listView removeFromSuperview];
    }
    [_listDict removeAllObjects];
    
    __weak __typeof(self) weakSelf = self;
    [self refreshWidthCompletion:^(CGFloat width) {
        __strong __typeof(self) self = weakSelf;
        
        [self.listCollectionView setContentOffset:CGPointMake(width * self.currentIndex, 0) animated:NO];
        [self.listCollectionView reloadData];
    }];
}

#pragma mark - UICollectionViewDataSource
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return self.isLoaded ? 1 : 0;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return [self.delegate numberOfListsInSmoothView:self];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:GKPageSmoothViewCellID forIndexPath:indexPath];
    id<GKPageSmoothListViewDelegate> list = self.listDict[@(indexPath.item)];
    if (list == nil) {
        list = [self.delegate smoothView:self initListAtIndex:indexPath.item];
        _listDict[@(indexPath.item)] = list;
        [list.listView setNeedsLayout];
        [list.listView layoutIfNeeded];
        
        UIScrollView *listScrollView = list.listScrollView;
        if ([listScrollView isKindOfClass:[UITableView class]]) {
            ((UITableView *)listScrollView).estimatedRowHeight = 0;
            ((UITableView *)listScrollView).estimatedSectionHeaderHeight = 0;
            ((UITableView *)listScrollView).estimatedSectionFooterHeight = 0;
        }
        if (@available(iOS 11.0, *)) {
            listScrollView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
        }
        listScrollView.contentInset = UIEdgeInsetsMake(self.headerContainerHeight, 0, 0, 0);
        if (CGSizeEqualToSize(listScrollView.contentSize, CGSizeZero)) {
            listScrollView.contentSize = CGSizeMake(listScrollView.contentSize.width, self.bounds.size.height);
        }
        
        self.currentListInitializeContentOffsetY = -listScrollView.contentInset.top + MIN(-self.currentHeaderContainerViewY, (self.headerHeight - self.ceilPointHeight));
        listScrollView.contentOffset = CGPointMake(0, self.currentListInitializeContentOffsetY);
        UIView *listHeader = [[UIView alloc] initWithFrame:CGRectMake(0, -self.headerContainerHeight, self.bounds.size.width, self.headerContainerHeight)];
        [listScrollView addSubview:listHeader];
        if (self.headerContainerView.superview == nil) {
            [listHeader addSubview:self.headerContainerView];
        }
        self.listHeaderDict[@(indexPath.item)] = listHeader;
        [listScrollView addObserver:self forKeyPath:@"contentOffset" options:NSKeyValueObservingOptionNew context:nil];
    }
    for (id<GKPageSmoothListViewDelegate> listItem in self.listDict.allValues) {
        listItem.listScrollView.scrollsToTop = (listItem == list);
    }
    
    UIView *listView = list.listView;
    if (listView != nil && listView.superview != cell.contentView) {
        for (UIView *view in cell.contentView.subviews) {
            [view removeFromSuperview];
        }
        listView.frame = cell.contentView.bounds;
        [cell.contentView addSubview:listView];
    }
    return cell;
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return self.bounds.size;
}

- (void)collectionView:(UICollectionView *)collectionView willDisplayCell:(UICollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath {
    [self listDidAppear:indexPath.item];
}

- (void)collectionView:(UICollectionView *)collectionView didEndDisplayingCell:(UICollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath {
    [self listDidDisappear:indexPath.item];
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    if ([self.delegate respondsToSelector:@selector(smoothView:scrollViewDidScroll:)]) {
        [self.delegate smoothView:self scrollViewDidScroll:scrollView];
    }
    
    NSInteger index = scrollView.contentOffset.x / scrollView.bounds.size.width;
    
    NSInteger ratio = (int)scrollView.contentOffset.x % (int)scrollView.bounds.size.width;
    
    UIScrollView *listScrollView = self.listDict[@(index)].listScrollView;
    if (index != self.currentIndex && ratio == 0 && !(scrollView.isDragging || scrollView.isDecelerating) && listScrollView.contentOffset.y <= -(self.segmentedHeight + self.ceilPointHeight)) {
        [self horizontalScrollDidEndAtIndex:index];
    }else {
        // 左右滚动的时候，把headerContainerView添加到self，达到悬浮的效果
        if (self.headerContainerView.superview != self) {
            self.headerContainerView.frame = CGRectMake(0, self.currentHeaderContainerViewY, self.headerContainerView.bounds.size.width, self.headerContainerView.bounds.size.height);
            [self addSubview:self.headerContainerView];
        }
    }
    if (self.currentIndex != index) {
        self.currentIndex = index;
    }
}

- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate {
    if (!decelerate) {
        NSInteger index = scrollView.contentOffset.x / scrollView.bounds.size.width;
        [self horizontalScrollDidEndAtIndex:index];
    }
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView {
    NSInteger index = scrollView.contentOffset.x / scrollView.bounds.size.width;
    [self horizontalScrollDidEndAtIndex:index];
}

#pragma mark - KVO
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if ([keyPath isEqualToString:@"contentOffset"]) {
        UIScrollView *scrollView = (UIScrollView *)object;
        if (scrollView != nil) {
            [self listScrollViewDidScroll:scrollView];
        }
    }else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

#pragma mark - Private Methods
- (void)listScrollViewDidScroll:(UIScrollView *)scrollView {
    if (self.listCollectionView.isDragging || self.listCollectionView.isDecelerating) return;
    
    NSInteger listIndex = [self listIndexForListScrollView:scrollView];
    if (listIndex != self.currentIndex) return;
    self.currentListScrollView = scrollView;
    CGFloat contentOffsetY = scrollView.contentOffset.y + self.headerContainerHeight;
    
    if (contentOffsetY < (self.headerHeight - self.ceilPointHeight)) {
        self.syncListContentOffsetEnabled = YES;
        self.currentHeaderContainerViewY = -contentOffsetY;
        for (id<GKPageSmoothListViewDelegate> list in self.listDict.allValues) {
            if (list.listScrollView != scrollView) {
                [list.listScrollView setContentOffset:scrollView.contentOffset animated:NO];
            }
        }
        UIView *listHeader = [self listHeaderForListScrollView:scrollView];
        if (self.headerContainerView.superview != listHeader) {
            self.headerContainerView.frame = CGRectMake(0, 0, self.headerContainerView.bounds.size.width, self.headerContainerView.bounds.size.height);
            [listHeader addSubview:self.headerContainerView];
        }
        
        if (self.isControlVerticalIndicator && self.ceilPointHeight != 0) {
            self.currentListScrollView.showsVerticalScrollIndicator = NO;
        }
    }else {
        if (self.headerContainerView.superview != self) {
            self.headerContainerView.frame = CGRectMake(0, -(self.headerHeight - self.ceilPointHeight), self.headerContainerView.bounds.size.width, self.headerContainerView.bounds.size.height);
            [self addSubview:self.headerContainerView];
        }
        
        if (self.isControlVerticalIndicator) {
            self.currentListScrollView.showsVerticalScrollIndicator = YES;
        }
        
        if (self.syncListContentOffsetEnabled) {
            self.syncListContentOffsetEnabled = NO;
            self.currentHeaderContainerViewY = -(self.headerHeight - self.ceilPointHeight);
            for (id<GKPageSmoothListViewDelegate> listItem in self.listDict.allValues) {
                if (listItem.listScrollView != scrollView) {
                    [listItem.listScrollView setContentOffset:CGPointMake(0, -(self.segmentedHeight + self.ceilPointHeight)) animated:NO];
                }
            }
        }
    }
}

- (void)refreshWidthCompletion:(void(^)(CGFloat width))completion {
    __block CGFloat width = self.bounds.size.width;
    
    if (width == 0) {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1f * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            width = self.bounds.size.width;
            !completion ? : completion(width);
        });
    }else {
        !completion ? : completion(width);
    }
}

- (void)horizontalScrollDidEndAtIndex:(NSInteger)index {
    self.currentIndex = index;
    UIView *listHeader = self.listHeaderDict[@(index)];
    UIScrollView *listScrollView = self.listDict[@(index)].listScrollView;
    if (listHeader != nil && listScrollView.contentOffset.y <= -(self.segmentedHeight + self.ceilPointHeight)) {
        for (id<GKPageSmoothListViewDelegate> listItem in self.listDict.allValues) {
            listItem.listScrollView.scrollsToTop = (listItem.listScrollView == listScrollView);
        }
        self.headerContainerView.frame = CGRectMake(0, 0, self.headerContainerView.bounds.size.width, self.headerContainerView.bounds.size.height);
        [listHeader addSubview:self.headerContainerView];
    }
}

- (UIView *)listHeaderForListScrollView:(UIScrollView *)scrollView {
    for (NSNumber *index in self.listDict) {
        if (self.listDict[index].listScrollView == scrollView) {
            return self.listHeaderDict[index];
        }
    }
    return nil;
}

- (NSInteger)listIndexForListScrollView:(UIScrollView *)scrollView {
    for (NSNumber *index in self.listDict) {
        if (self.listDict[index].listScrollView == scrollView) {
            return index.integerValue;
        }
    }
    return 0;
}

- (void)listDidAppear:(NSInteger)index {
    NSUInteger count = [self.delegate numberOfListsInSmoothView:self];
    if (count <= 0 || index >= count) {
        return;
    }
    
    id<GKPageSmoothListViewDelegate> list = self.listDict[@(index)];
    if (list && [list respondsToSelector:@selector(listViewDidAppear)]) {
        [list listViewDidAppear];
    }
}

- (void)listDidDisappear:(NSInteger)index {
    NSUInteger count = [self.delegate numberOfListsInSmoothView:self];
    if (count <= 0 || index >= count) {
        return;
    }
    
    id<GKPageSmoothListViewDelegate> list = self.listDict[@(index)];
    if (list && [list respondsToSelector:@selector(listViewDidDisappear)]) {
        [list listViewDidDisappear];
    }
}

@end

File: GKPageScrollView/swift/GKPageSmoothView.swift
//
//  GKPageSmoothView.swift
//  GKPageScrollViewSwift
//
//  Created by QuintGao on 2020/5/10.
//  Copyright © 2020 gaokun. All rights reserved.
//

import UIKit

@objc public protocol GKPageSmoothListViewDelegate : NSObjectProtocol {
    
    /// 返回listView，如果是vc就返回vc.view,如果是自定义view，就返回view本身
    func listView() -> UIView
    
    /// 返回vc或view内部持有的UIScrollView或UITableView或UICollectionView
    func listScrollView() -> UIScrollView
    
    @objc optional func listViewDidAppear()
    @objc optional func listViewDidDisappear()
}

@objc public protocol GKPageSmoothViewDelegate : NSObjectProtocol {
    
    /// 返回页面header视图
    /// - Parameter smoothView: smoothView description
    func headerView(in smoothView: GKPageSmoothView) -> UIView
    
    /// 返回页面分段视图
    /// - Parameter smoothView: smoothView description
    func segmentedView(in smoothView: GKPageSmoothView) -> UIView
    
    /// 返回个数列表
    /// - Parameter smoothView: smoothView description
    func numberOfLists(in smoothView: GKPageSmoothView) -> Int
    
    /// 根据index初始化一个列表实例，需实现`GKPageSmoothListViewDelegate`代理
    /// - Parameters:
    ///   - smoothView: smoothView description
    ///   - index: 对应的索引
    func smoothView(_ smoothView: GKPageSmoothView, initListAtIndex index: Int) -> GKPageSmoothListViewDelegate
    
    /// containerView滑动代理
    /// - Parameters:
    ///   - smoothView: smoothView description
    ///   - scrollView: containerScrollView
    @objc optional func smoothViewDidScroll(_ smoothView: GKPageSmoothView, scrollView: UIScrollView)
}

public class GKPageSmoothCollectionView: UICollectionView, UIGestureRecognizerDelegate {
    var headerContainerView: UIView?
    public func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
        let point = touch.location(in: headerContainerView)
        if headerContainerView?.bounds.contains(point) == true {
            return false
        }
        return true
    }
}

let GKPageSmoothViewCellID = "smoothViewCell"

open class GKPageSmoothView: UIView {
    public private(set) var listDict = [Int: GKPageSmoothListViewDelegate]()
    public let listCollectionView: GKPageSmoothCollectionView
    public var defaultSelectedIndex: Int = 0
    public var ceilPointHeight: CGFloat = GKPage_NavBar_Height
    public var isControlVerticalIndicator: Bool = false
    
    weak var delegate: GKPageSmoothViewDelegate?
    var listHeaderDict = [Int: UIView]()
    let headerContainerView: UIView
    var currentListScrollView: UIScrollView?
    var isSyncListContentOffsetEnabled: Bool = false
    var currentHeaderContainerViewY: CGFloat = 0
    
    var headerContainerHeight: CGFloat = 0
    var headerHeight: CGFloat = 0
    var segmentedHeight: CGFloat = 0
    var currentListInitailzeContentOffsetY: CGFloat = 0
    
    var currentIndex: Int = 0
    var isLoaded: Bool = false
    var isRefreshHeader: Bool = false

    deinit {
        listDict.values.forEach {
            $0.listScrollView().removeObserver(self, forKeyPath: "contentOffset")
            $0.listScrollView().removeObserver(self, forKeyPath: "contentSize")
        }
    }
    
    public init(delegate: GKPageSmoothViewDelegate) {
        self.delegate = delegate
        
        headerContainerView = UIView()
        
        let layout = UICollectionViewFlowLayout()
        layout.minimumLineSpacing = 0
        layout.minimumInteritemSpacing = 0
        layout.scrollDirection = .horizontal
        listCollectionView = GKPageSmoothCollectionView(frame: .zero, collectionViewLayout: layout)
        super.init(frame: .zero)
        
        listCollectionView.dataSource = self
        listCollectionView.delegate = self
        listCollectionView.isPagingEnabled = true
        listCollectionView.bounces = false
        listCollectionView.showsHorizontalScrollIndicator = false
        listCollectionView.scrollsToTop = false
        listCollectionView.register(UICollectionViewCell.classForCoder(), forCellWithReuseIdentifier: GKPageSmoothViewCellID)
        if #available(iOS 10.0, *) {
            listCollectionView.isPrefetchingEnabled = false
        }
        if #available(iOS 11.0, *) {
            listCollectionView.contentInsetAdjustmentBehavior = .never
        }
        listCollectionView.headerContainerView = headerContainerView
        self.addSubview(listCollectionView)
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public func refreshHeaderView() {
        isRefreshHeader = true
        reloadData()
    }
    
    public func reloadData() {
        guard let delegate = delegate else { return }
        currentListScrollView = nil
        if isRefreshHeader {
            isRefreshHeader = false
        }else {
            currentIndex = defaultSelectedIndex
        }
        currentHeaderContainerViewY = 0
        isSyncListContentOffsetEnabled = false
        isLoaded = true
        
        listHeaderDict.removeAll()
        listDict.values.forEach {
            $0.listScrollView().removeObserver(self, forKeyPath: "contentOffset")
            $0.listScrollView().removeObserver(self, forKeyPath: "contentSize")
            $0.listView().removeFromSuperview()
        }
        listDict.removeAll()
        
        let headerView = delegate.headerView(in: self)
        let segmentedView = delegate.segmentedView(in: self)
        headerContainerView.addSubview(headerView)
        headerContainerView.addSubview(segmentedView)
        
        headerHeight = headerView.bounds.size.height
        segmentedHeight = segmentedView.bounds.size.height
        headerContainerHeight = headerHeight + segmentedHeight
        
        var width: CGFloat = bounds.size.width
        if width == 0 {
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.0) {
                width = self.bounds.size.width
                
                self.headerContainerView.frame = CGRect(x: 0, y: 0, width: width, height: self.headerContainerHeight)
                headerView.frame = CGRect(x: 0, y: 0, width: width, height: self.headerHeight)
                segmentedView.frame = CGRect(x: 0, y: self.headerHeight, width: width, height: self.segmentedHeight)
                self.listCollectionView.setContentOffset(CGPoint(x: width * CGFloat(self.currentIndex), y: 0), animated: false)
                self.listCollectionView.reloadData()
            }
        }else {
            self.headerContainerView.frame = CGRect(x: 0, y: 0, width: width, height: self.headerContainerHeight)
            headerView.frame = CGRect(x: 0, y: 0, width: width, height: self.headerHeight)
            segmentedView.frame = CGRect(x: 0, y: self.headerHeight, width: width, height: self.segmentedHeight)
            self.listCollectionView.setContentOffset(CGPoint(x: width * CGFloat(self.currentIndex), y: 0), animated: false)
            self.listCollectionView.reloadData()
        }
    }
    
    open override func layoutSubviews() {
        super.layoutSubviews()
        
        listCollectionView.frame = bounds
//        if headerContainerView.frame == .zero {
//            reloadData()
//        }
    }
    
    // MARK: - KVO
    open override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if keyPath == "contentOffset" {
            if let scrollView = object as? UIScrollView {
                listDidScroll(scrollView: scrollView)
            }
        }else if keyPath == "contentSize" {
            if let scrollView = object as? UIScrollView {
                let minContentSizeHeight = bounds.size.height - (self.headerHeight - self.ceilPointHeight)
                if minContentSizeHeight > scrollView.contentSize.height {
                    scrollView.contentSize = CGSize(width: scrollView.contentSize.width, height: minContentSizeHeight)
                    scrollView.contentOffset = CGPoint(x: 0, y: self.currentHeaderContainerViewY)
                }
            }
        }else {
            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
        }
    }
    
    // MARK: - Private Methods
    func listDidScroll(scrollView: UIScrollView) {
        if listCollectionView.isDragging || listCollectionView.isDecelerating {
            return
        }
        let index = listIndex(for: scrollView)
        if index != currentIndex {
            return
        }
        currentListScrollView = scrollView
        let contentOffsetY = scrollView.contentOffset.y + headerContainerHeight
        if contentOffsetY < (headerHeight - ceilPointHeight) {
            isSyncListContentOffsetEnabled = true
            currentHeaderContainerViewY = -contentOffsetY
            for list in listDict.values {
                if list.listScrollView() != scrollView {
                    list.listScrollView().setContentOffset(scrollView.contentOffset, animated: false)
                }
            }
            let header = listHeader(for: scrollView)
            if headerContainerView.superview != header {
                headerContainerView.frame.origin.y = 0
                header?.addSubview(headerContainerView)
            }
            
            if isControlVerticalIndicator && ceilPointHeight != 0 {
                currentListScrollView?.showsVerticalScrollIndicator = false
            }
        }else {
            if headerContainerView.superview != self {
                headerContainerView.frame.origin.y = -(headerHeight - ceilPointHeight)
                addSubview(headerContainerView)
            }
            
            if isControlVerticalIndicator {
                currentListScrollView?.showsVerticalScrollIndicator = true
            }
            
            if isSyncListContentOffsetEnabled {
                isSyncListContentOffsetEnabled = false
                currentHeaderContainerViewY = -(headerHeight - ceilPointHeight)
                for list in listDict.values {
                    if list.listScrollView() != currentListScrollView {
                        list.listScrollView().setContentOffset(CGPoint(x: 0, y: -(segmentedHeight + ceilPointHeight)), animated: false)
                    }
                }
            }
        }
    }
    
    func horizontalScrollDidEnd(at index: Int) {
        currentIndex = index
        guard let listHeader = listHeaderDict[index], let listScrollView = listDict[index]?.listScrollView() else { return }
        listDict.values.forEach {
            $0.listScrollView().scrollsToTop = ($0.listScrollView() == listScrollView)
        }
        if listScrollView.contentOffset.y <= -(segmentedHeight + ceilPointHeight) {
            headerContainerView.frame.origin.y = 0
            listHeader.addSubview(headerContainerView)
        }
    }
    
    func listHeader(for listScrollView: UIScrollView) -> UIView? {
        for (index, list) in listDict {
            if list.listScrollView() == listScrollView {
                return listHeaderDict[index]
            }
        }
        return nil
    }
    
    func listIndex(for listScrollView: UIScrollView) -> Int {
        for (index, list) in listDict {
            if list.listScrollView() == listScrollView {
                return index
            }
        }
        return 0
    }
    
    func listDidAppear(at index: Int) {
        guard let delegate = delegate else { return }
        let count = delegate.numberOfLists(in: self)
        if count <= 0 || index >= count {
            return
        }
        listDict[index]?.listViewDidAppear?()
    }
    
    func listDidDisappear(at index: Int) {
        guard let delegate = delegate else { return }
        let count = delegate.numberOfLists(in: self)
        if count <= 0 || index >= count {
            return
        }
        listDict[index]?.listViewDidDisappear?()
    }
}

extension GKPageSmoothView: UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        guard let delegate = delegate else { return 0 }
        
        let count = delegate.numberOfLists(in: self)
        
        return self.isLoaded ? count : 0
    }
    
    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        guard let delegate = delegate else { return UICollectionViewCell() }
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: GKPageSmoothViewCellID, for: indexPath)
        var list = listDict[indexPath.item]
        if list == nil {
            list = delegate.smoothView(self, initListAtIndex: indexPath.item)
            listDict[indexPath.item] = list!
            list?.listView().setNeedsLayout()
            list?.listView().layoutIfNeeded()
            if list?.listScrollView().isKind(of: UITableView.self) == true {
                (list?.listScrollView() as? UITableView)?.estimatedRowHeight = 0
                (list?.listScrollView() as? UITableView)?.estimatedSectionHeaderHeight = 0
                (list?.listScrollView() as? UITableView)?.estimatedSectionFooterHeight = 0
            }
            if #available(iOS 11.0, *) {
                list?.listScrollView().contentInsetAdjustmentBehavior = .never
            }
            list?.listScrollView().contentInset = UIEdgeInsets(top: headerContainerHeight, left: 0, bottom: 0, right: 0)
            currentListInitailzeContentOffsetY = -headerContainerHeight + min(-currentHeaderContainerViewY, (headerHeight - ceilPointHeight))
            list?.listScrollView().contentOffset = CGPoint(x: 0, y: currentListInitailzeContentOffsetY)
            let listHeader = UIView(frame: CGRect(x: 0, y: -headerContainerHeight, width: bounds.size.width, height: headerContainerHeight))
            list?.listScrollView().addSubview(listHeader)
            if headerContainerView.superview == nil {
                listHeader.addSubview(headerContainerView)
            }
            listHeaderDict[indexPath.item] = listHeader
            list?.listScrollView().addObserver(self, forKeyPath: "contentOffset", options: .new, context: nil)
            list?.listScrollView().addObserver(self, forKeyPath: "contentSize", options: .new, context: nil)
        }
        listDict.values.forEach {
            $0.listScrollView().scrollsToTop = ($0 === list)
        }
        if let listView = list?.listView(), listView.superview != cell.contentView {
            cell.contentView.subviews.forEach { $0.removeFromSuperview() }
            listView.frame = cell.contentView.bounds
            cell.contentView.addSubview(listView)
        }
        return cell
    }
    
    public func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return self.bounds.size
    }
    
    public func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
        listDidAppear(at: indexPath.item)
    }
    
    public func collectionView(_ collectionView: UICollectionView, didEndDisplaying cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
        listDidDisappear(at: indexPath.item)
    }
    
    public func scrollViewDidScroll(_ scrollView: UIScrollView) {
        delegate?.smoothViewDidScroll?(self, scrollView: scrollView)
        let index = Int(scrollView.contentOffset.x/scrollView.bounds.size.width)
        let ratio = Int(scrollView.contentOffset.x)%Int(scrollView.bounds.size.width)
        
        let listScrollView = self.listDict[index]?.listScrollView()
        if index != currentIndex && ratio == 0 && !(scrollView.isDragging || scrollView.isDecelerating) && listScrollView?.contentOffset.y ?? 0 <= -(segmentedHeight + ceilPointHeight) {
            horizontalScrollDidEnd(at: index)
        }else {
            if headerContainerView.superview != self {
                headerContainerView.frame.origin.y = currentHeaderContainerViewY
                addSubview(headerContainerView)
            }
        }
        if currentIndex != index {
            currentIndex = index
        }
    }
    
    public func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        if !decelerate {
            let index = Int(scrollView.contentOffset.x / scrollView.bounds.size.width)
            horizontalScrollDidEnd(at: index)
        }
    }
    
    public func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        let index = Int(scrollView.contentOffset.x / scrollView.bounds.size.width)
        horizontalScrollDidEnd(at: index)
    }
}
